src/ethereum/frontier/vm/gas.py

```diff
@@ -45,3 +48,56 @@ def subtract_gas(gas_left: U256, amount: U256) -> U256:
         raise OutOfGasError
 
     return gas_left - amount
+
+
+def calculate_memory_gas_cost(size_in_bytes: U256) -> U256:
+    """
+    Calculates the gas cost for allocating memory up
+    and including the given address.
+
+    Parameters
+    ----------
+    size_in_bytes :
+        The size of the data in bytes.
+
+    Returns
+    -------
+    total_gas_cost : `ethereum.base_types.U256`
+        The gas cost for storing data in memory.
+    """
+    size_in_words = ceil32(size_in_bytes) // 32
+    linear_cost = size_in_words * GAS_MEMORY
+    quadratic_cost = size_in_words ** 2 // GAS_MEMORY_QUADRATIC_DENOMINATOR
+    total_gas_cost = linear_cost + quadratic_cost
+    return total_gas_cost
+
+
+def calculate_gas_for_extending_memory(
+    memory: bytearray, start_position: U256, size: U256
+) -> U256:
+    """
+    Calculates gas needed for extending memory.
+
+    Parameters
+    ----------
+    memory :
+        Memory contents of the EVM.
+
+    start_position :
+        Starting position of the memory.
+
+    size :
+        Size to extend memory.
+
+    Returns
+    -------
+    total_gas_cost : `ethereum.base_types.U256`
+        The gas cost for extending memory.
+    """
+    memory_size = U256(len(memory))
+    before_size = ceil32(memory_size)
+    after_size = ceil32(start_position + size)
```

Comment:
Not sure if you'd agree, but maybe an easier way of implementing this would be something like:

```suggestion
    after_size = ceil32(start_position + size)
    
    if after_size <= before_size:
        return U256(0)
```

Reply:
Please take a look at 850f028 and see if you'd be okay with that. I decided to club all of the logic into a single method and called it `extend_memory_and_subtract_gas`. I can change it if you don't like it.